spring-boot-project/spring-boot/src/main/java/org/springframework/boot/webservices/client/WebServiceTemplateBuilder.java

```diff
@@ -0,0 +1,1106 @@
+	 * @see WebServiceTemplateBuilder#build()
+	 * @see #configure(WebServiceTemplate)
+	 */
+
+	public <T extends WebServiceTemplate> T build(Class<T> webServiceTemplateClass) {
+		Assert.notNull(webServiceTemplateClass,
+				"webServiceTemplateClass must not be null");
+		return configure(BeanUtils.instantiateClass(webServiceTemplateClass));
+	}
+
+	/**
+	 * Configure the provided {@link WebServiceTemplate} instance using this builder.
+	 * @param <T> the type of web service template
+	 * @param webServiceTemplate the {@link WebServiceTemplate} to configure
+	 * @return the web service template instance
+	 * @see #build()
+	 * @see #build(Class)
+	 */
+	public <T extends WebServiceTemplate> T configure(T webServiceTemplate) {
+		Assert.notNull(webServiceTemplate, "webServiceTemplate must not be null");
+
+		configureSenders(webServiceTemplate);
+
+		if (!CollectionUtils.isEmpty(this.internalCustomizers)) {
+			for (WebServiceTemplateCustomizer internalCustomizer : this.internalCustomizers) {
+				internalCustomizer.customize(webServiceTemplate);
+			}
+		}
+
+		if (this.marshaller != null) {
+			webServiceTemplate.setMarshaller(this.marshaller);
+		}
+
+		if (this.unmarshaller != null) {
+			webServiceTemplate.setUnmarshaller(this.unmarshaller);
+		}
+
+		if (this.destinationProvider != null) {
+			webServiceTemplate.setDestinationProvider(this.destinationProvider);
+		}
+
+		if (this.transformerFactoryClass != null) {
+			webServiceTemplate.setTransformerFactoryClass(this.transformerFactoryClass);
+		}
+
+		if (this.messageFactory != null) {
+			webServiceTemplate.setMessageFactory(this.messageFactory);
+		}
+
+		if (!CollectionUtils.isEmpty(this.customizers)) {
+			for (WebServiceTemplateCustomizer customizer : this.customizers) {
+				customizer.customize(webServiceTemplate);
+			}
+		}
+
+		if (!CollectionUtils.isEmpty(this.interceptors)) {
+			webServiceTemplate.setInterceptors(
+					append(this.interceptors, webServiceTemplate.getInterceptors())
+							.toArray(new ClientInterceptor[0]));
+		}
+
+		return webServiceTemplate;
+	}
+
+	private <T extends WebServiceTemplate> void configureSenders(T webServiceTemplate) {
+
+		if (!CollectionUtils.isEmpty(this.webServiceMessageSenderSuppliers)) {
+			Set<WebServiceMessageSender> webServiceMessageSenders = new LinkedHashSet<>();
+			for (Supplier<? extends WebServiceMessageSender> webServiceMessageSenderSupplier : this.webServiceMessageSenderSuppliers) {
+				webServiceMessageSenders.add(webServiceMessageSenderSupplier.get());
+			}
+			webServiceTemplate.setMessageSenders(
+					webServiceMessageSenders.toArray(new WebServiceMessageSender[0]));
+		}
+		else if (this.detectWebServiceMessageSender) {
+			webServiceTemplate.setMessageSenders(
+					new WebServiceMessageSender[] { detectMessageSender() });
+		}
+
+		if (!CollectionUtils.isEmpty(this.webServiceMessageSenderCustomizers)) {
+			if (!ObjectUtils.isEmpty(webServiceTemplate.getMessageSenders())) {
```

Comment:
This looks weird to me. Why do we check if the template has existing messageSenders?

Reply:
You can easily override a `WebServiceTemplate` and override `default strategies`.

```
    protected void initDefaultStrategies() {
        DefaultStrategiesHelper strategiesHelper = new DefaultStrategiesHelper(WebServiceTemplate.class);
        if (this.getMessageFactory() == null) {
            this.initMessageFactory(strategiesHelper);
        }

        if (ObjectUtils.isEmpty(this.getMessageSenders())) {
            this.initMessageSenders(strategiesHelper);
        }

        if (this.getFaultMessageResolver() == null) {
            this.initFaultMessageResolver(strategiesHelper);
        }

    }

```